distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso

```diff
@@ -0,0 +1,515 @@
+       be specified to set the precision and `Rounding_Mode`.
+
+       Arguments:
+       - that: The number to divide by this.
+       - math_context: Used to optionally specify precision and `Rounding_Mode`.
+
+       Division in Enso will undergo automatic conversions such that you need
+       not convert other numeric types to `Decimal` manually.
+
+       ! Error Conditions
+
+         - If the precision specified in `math_context` is 0, or if
+           `match_context` is unspecified, and the quotient has a nonterminating
+           decimal expansion, an `Arithmetic_Error` is thrown.
+
+       > Example
+         Dividing 1065.9378 by 23.34.
+
+             a = Decimal.new "1065.9378"
+             b = Decimal.new "23.34"
+             c = a.divide b
+             # => Decimal.new 45.67
+
+       > Example
+         Dividing 1065.9378 by 23.34, rounding to 3 digits of precision (1
+         decimal place), and rounding half-down.
+
+            a = Decimal.new "1065.9378"
+             b = Decimal.new "23.34"
+             a.divide b (Math_Context.new 3 Rounding_Mode.half_down)
+             # => Decimal.new 45.7
+    divide : Decimal -> Decimal -> Math_Context | Nothing -> Decimal | Arithmetic_Error
+    divide self (that:Decimal) (math_context : Math_Context | Nothing)=Nothing -> Decimal ! Arithmetic_Error =
+        handle_java_exception <|
+            case math_context of
+                Nothing -> Decimal.Value (self.big_decimal.divide that.big_decimal)
+                _ -> Decimal.Value (self.big_decimal.divide that.big_decimal math_context.math_context)
+
+    ## GROUP Operators
+       ICON math
+       Divides a `Decimal` by another `Decimal` or other kind of number, or
+       divides another kind of number by a `Decimal`.
+
+       Arguments:
+       - that: The number to divide by this.
+
+       Division in Enso will undergo automatic conversions such that you need
+       not convert other numeric types to `Decimal` manually.
+
+       > Example
+         Dividing 1065.9378 by 23.34.
+
+             a = Decimal.new "1065.9378"
+             b = Decimal.new "23.34"
+             c = a / b
+             # => Decimal.new 45.67
+    / : Decimal -> Decimal -> Decimal
+    / self (that:Decimal) = self.divide that
+
+    ## GROUP Operators
+       Compute the negation of this.
+
+       > Example
+         Negate 5.1 to get -5.1.
+
+             5.1.negate
+             # => Decimal.new -5.1
+    negate : Decimal
+    negate self = Decimal.Value self.big_decimal.negate
+
+    # PRIVATE
+    get_precision : Integer
+    get_precision self = self.big_decimal.precision
+
+    # PRIVATE
+    get_scale : Integer
+    get_scale self = self.big_decimal.scale
+
+    # PRIVATE
+    get_unscaled_value : Integer
+    # TODO: This +0 is needed to allow internal represeantations to be compared.
+    get_unscaled_value self = self.big_decimal.unscaledValue + 0
```

Comment:
@JaroslavTulach `self.big_decimal.unscaledValue` here returns a BigInteger directly, which I assume is the wrong thing to do (especially since it is often small enough to fit in Long). I added `+ 0` because if I don't, I am unable to compare arrays containing this value with `==`. What is the right way to return a BigInteger here?

Reply:
> Shall we have ToEnsoNumberNode in HostValueToEnsoNode for BigIntegers?

Possibly, but it may clash with _preserving identity_ requested by Truffle TCK.

> unable to compare arrays containing this value with ==

That's certainly a bug.

Maybe: Integrate as it is and report a bug requesting to modify this line to simple `self.big_decimal.unscaledValue` engine.

Reply:
#9392